lib/vector-core/buffers/benches/common.rs

```diff
@@ -0,0 +1,177 @@
+use buffers::bytes::{DecodeBytes, EncodeBytes};
+use buffers::{self, Variant};
+use bytes::{Buf, BufMut};
+use futures::task::{noop_waker, Context, Poll};
+use futures::{Sink, Stream};
+use std::fmt;
+use std::pin::Pin;
+
+#[derive(Clone, Copy)]
+pub struct Message<const N: usize> {
+    id: u64,
+    _padding: [u64; N],
+}
+
+impl<const N: usize> Message<N> {
+    fn new(id: u64) -> Self {
+        Message {
+            id,
+            _padding: [0; N],
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum EncodeError {}
+
+#[derive(Debug)]
+pub enum DecodeError {}
+
+impl fmt::Display for DecodeError {
+    fn fmt(&self, _f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        unreachable!()
+    }
+}
+
+impl<const N: usize> EncodeBytes<Message<N>> for Message<N> {
+    type Error = EncodeError;
+
+    fn encode<B>(self, buffer: &mut B) -> Result<(), Self::Error>
+    where
+        B: BufMut,
+        Self: Sized,
+    {
+        buffer.put_u64(self.id);
+        for _ in 0..N {
+            // this covers self._padding
+            buffer.put_u64(0);
+        }
+        Ok(())
+    }
+}
+
+impl<const N: usize> DecodeBytes<Message<N>> for Message<N> {
+    type Error = DecodeError;
+
+    fn decode<B>(mut buffer: B) -> Result<Self, Self::Error>
+    where
+        B: Buf,
+        Self: Sized,
+    {
+        let id = buffer.get_u64();
+        for _ in 0..N {
+            // this covers self._padding
+            let _ = buffer.get_u64();
+        }
+        Ok(Message::new(id))
+    }
+}
+
+#[allow(clippy::type_complexity)]
+pub fn setup<const N: usize>(
+    max_events: usize,
+    variant: Variant,
+) -> (
+    Box<dyn Sink<Message<N>, Error = ()> + Unpin + Send>,
+    Box<dyn Stream<Item = Message<N>> + Unpin + Send>,
```

Comment:
While we're on the subject, I'd also like to try and bend the project to use static dispatch instead of dynamic, but I figured putting in the effort there would run into serious roadblocks, be better sequenced after we have more of core extracted and be easier to pull off with better performance debug tools on hand.